src/LanguageServer/Protocol/Extensions/SourceGeneratedDocumentUri.cs

```diff
@@ -0,0 +1,95 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Specialized;
+using System.Linq;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer;
+
+// For source generated documents, we'll produce a URI specifically for LSP that has a scheme the client can register for; the "host" portion will
+// just be the project ID of the document, and the path will be the hint text for the document. This recognizes that VS Code shows just the local
+// path portion in the UI and thus embedding more into the path will appear and we don't want that. The rest of the stuff to serialize, namely the DocumentId, and any information
+// for the SourceGeneratedDocumentIdentity are put in as query string arguments
+//
+// For example, the URI can look like:
+//
+// roslyn-source-generated://E7D5BCFA-E345-4029-9D12-3EDCD0FB0F6B/Generated.cs?documentId=8E4C0B71-4044-4247-BDD0-04AF4C9E1677&assembly=Generator...
+//
+// where the first GUID is the project ID, the second GUID is the document ID.
+internal static class SourceGeneratedDocumentUri
+{
+    public const string Scheme = "roslyn-source-generated";
+    private const string GuidFormat = "D";
+    private const string DocumentIdParam = "documentId";
+    private const string HintNameParam = "hintName";
+    private const string AssemblyNameParam = "assemblyName";
+    private const string AssemblyVersionParam = "assemblyVersion";
+    private const string AssemblyPathParam = "assemblyPath";
+    private const string TypeNameParam = "typeName";
+
+    public static Uri Create(SourceGeneratedDocumentIdentity identity)
+    {
+        // Ensure the hint path is converted to a URI-friendly format
+        var hintPathParts = identity.HintName.Split('\\');
+        var hintPathPortion = string.Join("/", hintPathParts.Select(Uri.EscapeDataString));
+
+        var projectId = identity.DocumentId.ProjectId.Id.ToString(GuidFormat);
+        var documentId = identity.DocumentId.Id.ToString(GuidFormat);
+        var hintName = Uri.EscapeDataString(identity.HintName);
+        var assemblyName = Uri.EscapeDataString(identity.Generator.AssemblyName);
+        var assemblyVersion = Uri.EscapeDataString(identity.Generator.AssemblyVersion.ToString());
+        var typeName = Uri.EscapeDataString(identity.Generator.TypeName);
+
+        var uri = $"{Scheme}://{projectId}/{hintPathPortion}?{DocumentIdParam}={documentId}&{HintNameParam}={hintName}&{AssemblyNameParam}={assemblyName}&{AssemblyVersionParam}={assemblyVersion}&{TypeNameParam}={typeName}";
```

Comment:
so, the part i trust is the projectId/docId.  teh rest i think should all be opaque data we effectively don't munge with (specifically referring to hintpathportion).